crates/lace-bin/src/typechecking/check_pass/check.rs

```diff
@@ -0,0 +1,564 @@
+use crate::ast_metadata::{Metadata, MetadataId};
+use crate::lice::Lice;
+use crate::parser::ast::{
+    BinaryOp, Block, Expr, ExprKind, Function, Item, Lit, Statement, UnaryOp,
+};
+use crate::typechecking::ctx::TypeContext;
+use crate::typechecking::error::{FailedUnification, ResultExt, TypeError};
+use crate::typechecking::ty::PartialType;
+use crate::typechecking::type_spec_to_partial_type;
+use bumpalo::collections::Vec as BumpVec;
+
+#[derive(Copy, Clone, Debug)]
+pub struct ReturnContext<'a> {
+    // the expected outcome of the current expr
+    expected_type: PartialType<'a>,
+
+    // for explicit returns, return expressions
+    #[allow(unused)]
+    function_return_type: PartialType<'a>,
+    function_return_type_span: MetadataId,
+
+    // for breaks and return expressions etc.
+    block_return_type: PartialType<'a>,
+
+    // is the block we're typechecking the toplevel scope of a function?
+    function_block: bool,
+}
+
+fn typecheck_lit<'a>(
+    lit: &Metadata<&Lit>,
+    ctx: &mut TypeContext<'a, '_, '_>,
+    rctx: ReturnContext<'a>,
+) -> Result<(), TypeError> {
+    let res = match lit.value {
+        Lit::Bool(_) => ctx.unify(PartialType::Bool, rctx.expected_type),
+        Lit::Int(_) => ctx.unify(PartialType::Int, rctx.expected_type),
+        Lit::String(_) => ctx.unify(PartialType::String, rctx.expected_type),
+    };
+
+    if let Err((l, r)) = res {
+        return Err(TypeError::FailedUnification(FailedUnification {
+            expected: r.to_string(),
+            was: l.to_string(),
+            was_span: ctx.span_for(lit.metadata),
+        }));
+    }
+
+    Ok(())
+}
+
+fn typecheck_expr<'a>(
+    expr: &Expr,
+    ctx: &mut TypeContext<'a, '_, '_>,
+    rctx: ReturnContext<'a>,
+) -> Result<(), TypeError> {
+    match &expr.value {
+        ExprKind::Lit(lit) => typecheck_lit(
+            &Metadata {
+                value: lit,
+                metadata: expr.metadata,
+            },
+            ctx,
+            rctx,
+        ),
+        ExprKind::If(condition, l, r) => {
+            typecheck_expr(
+                condition,
+                ctx,
+                ReturnContext {
+                    expected_type: PartialType::Bool,
+                    ..rctx
+                },
+            )
+            .on_failed_unification(|uni| TypeError::IfCondition {
+                condition_span: ctx.span_for(condition.metadata),
+                condition_ty: uni.was,
+                if_span: ctx.span_for(expr.metadata),
+            })?;
+
+            if let Some(r) = r {
+                let lvar = ctx.fresh();
+                let rvar = ctx.fresh();
+
+                typecheck_block(
+                    l,
+                    ctx,
+                    &ReturnContext {
+                        expected_type: lvar.into(),
+                        function_block: false,
+                        ..rctx
+                    },
```

Comment:
So it feels like this context is not useful apart from error messages? It feels like you're adding information into every call instead of transforming errors, but maybe this is better. Does this come from literature?